File: app/src/main/java/com/mailerdaemon/app/ImpContacts/ContactDetailFragment.java
package com.mailerdaemon.app.ImpContacts;

import android.annotation.SuppressLint;
import android.app.SearchManager;
import android.content.Context;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.SearchView;

import com.google.gson.Gson;
import com.mailerdaemon.app.R;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

import Utils.SearchObservabel;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

public class ContactDetailFragment extends Fragment {

    private RecyclerView recyclerView;
    private List<Contact> contactList;
    private ContactRecyclerAdapter adapter;
    SearchView searchView;

    @SuppressLint("CheckResult")
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        contactList= new Gson().fromJson(loadJSONFromAsset(getArguments().getString("type")), FacultyModel.class).getContact();
        View view=inflater.inflate(R.layout.fragment_contact_detail,container,false);
        recyclerView=view.findViewById(R.id.rv_contact_detail);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        adapter=new ContactRecyclerAdapter();
        recyclerView.setAdapter(adapter);
        setHasOptionsMenu(true);
        //searchView=view.findViewById(R.id.search_view);
        //getActivity().setActionBar(view.findViewById(R.id.toolbar));


        return view;
    }
    private Observable<List<Contact>> dataFromNetwork(final String query) {
        if(query.isEmpty())
            return Observable.just(contactList);
        else
            return Observable.just(contactList).flatMap(Observable::fromIterable).
                    filter(pojo-> pojo.getName().contains(query))
                    .toList().toObservable();
    }

    private void updateRV(List<Contact> result) {
        DiffUtilCallBack n=new DiffUtilCallBack(adapter.getData(),result);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(n);
        adapter.setData(result);
        diffResult.dispatchUpdatesTo(adapter);
    }

    public String loadJSONFromAsset(String type) {
        String json = null;
        try {
            InputStream is = getActivity().getAssets().open(type+".json");
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            json = new String(buffer, "UTF-8");
        } catch (IOException ex) {
            ex.printStackTrace();
            return null;
        }
        return json;
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.imp_contact_search, menu);
        SearchManager searchManager =
                (SearchManager) getActivity().getSystemService(Context.SEARCH_SERVICE);
        searchView =
                (SearchView) menu.findItem(R.id.search).getActionView();
        menu.findItem(R.id.search).expandActionView();
        searchView.setQueryHint("Search Contacts");
        SearchObservabel.fromView(searchView)
                .debounce(300, TimeUnit.MILLISECONDS)
                .distinctUntilChanged()
                .switchMap((Function<String, ObservableSource<List<Contact>>>) this::dataFromNetwork)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(result -> {updateRV(result);});
        searchView.setSearchableInfo(
                searchManager.getSearchableInfo(getActivity().getComponentName()));
        super.onCreateOptionsMenu(menu,inflater);

    }


}

File: app/src/main/java/com/mailerdaemon/app/impContacts/ContactDetailFragment.kt
package com.mailerdaemon.app.impContacts

import android.annotation.SuppressLint
import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DiffUtil
import com.google.gson.Gson
import com.mailerdaemon.app.R
import com.mailerdaemon.app.toast
import com.mailerdaemon.app.utils.ContactFunction
import kotlinx.android.synthetic.main.fragment_contact_detail.*
import kotlinx.android.synthetic.main.fragment_contact_detail.view.*
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.ArrayList

class ContactDetailFragment : Fragment(), ContactFunction {
    private var contactList = ArrayList<Contact>()
    private val adapter: ContactRecyclerAdapter = ContactRecyclerAdapter(this)

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        contactList = (Gson().fromJson(loadJSONFromAsset(arguments
                ?.getString("type")), FacultyModel::class.java)
                .contact as? ArrayList<Contact> ?: ArrayList())
        val view = inflater.inflate(R.layout.fragment_contact_detail, container, false)
        view.rv_contact_detail.adapter = adapter
        adapter.data = contactList
        adapter.notifyDataSetChanged()
        setHasOptionsMenu(true)
        return view
    }

    private fun updateRV(result: List<Contact>) {
        val diffResult = DiffUtil.calculateDiff(DiffUtilCallBack(adapter.data, result))
        adapter.data = result
        diffResult.dispatchUpdatesTo(adapter)
    }

    private fun loadJSONFromAsset(type: String?): String? {
        val json: String
        try {
            val inputStream = (activity)?.assets?.open("$type.json")
            val size = inputStream?.available()
            val buffer = size?.let { ByteArray(it) }
            inputStream?.let {
                it.read(buffer)
                it.close()
            }
            json = buffer?.let { String(it, StandardCharsets.UTF_8) }.toString()
        } catch (ex: IOException) {
            ex.printStackTrace()
            return null
        }
        return json
    }

    @SuppressLint("CheckResult")
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.imp_contact_search, menu)
        val searchManager = (activity)?.getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.search).actionView as SearchView
        searchView.let {
            it.queryHint = "Search Contacts"
            it.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    rv_contact_detail.scrollToPosition(0)
                    searchView.clearFocus()
                    return true
                }
                override fun onQueryTextChange(newText: String): Boolean {
                    if (newText == "") {
                        updateRV(contactList)
                    } else {
                        val list = ArrayList<Contact>()
                        for (c in contactList) {
                            if (c.name?.contains(newText.toLowerCase()) == true)
                                list.add(c)
                        }
                        updateRV(list)
                    }
                    return true
                }
            })
            it.setOnFocusChangeListener { v, hasFocus ->
                val manager = v.context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
                if (hasFocus) manager.showSoftInput(v.findFocus(), InputMethodManager.RESULT_SHOWN) }
            it.setSearchableInfo(searchManager.getSearchableInfo(activity?.componentName))
            it.setIconifiedByDefault(false)
        }
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun makeCall(num: String) {
        if (num.trim() != "0") {
            val intent = Intent(Intent.ACTION_DIAL, Uri.parse("tel:" + num.trim()))
            startActivity(intent)
        } else context?.toast(resources.getString(R.string.number_not_available))
    }

    override fun sendMail(s: String) {
        if (s.trim().isNotEmpty()) {
            val send = Intent(Intent.ACTION_SENDTO)
            val uriText = "mailto:${Uri.encode(s.trim())}" +
                    "?subject=${Uri.encode("Subject")}" +
                    "&body=${Uri.encode("the body of the message")}"
            send.data = Uri.parse(uriText)
            startActivity(Intent.createChooser(send, "Send mail..."))
        } else context?.toast(resources.getString(R.string.email_not_available))
    }
}
